Fix trigger for f_sticky

complacentsee opened the pull request:
Bug fix to eliminate adding duplicate entries into f_sticky on replies.
While fixing I noticed a naming convention gap between the db schema migration and the new forum creation method. This cleans up both potential naming schemes and creates a new trigger that doesn't add duplicate records into f_sticky.

nyetwurk commented:
The migration should probably remove the duplicate entries in the various sticky tables (I can't see that it does so)

complacentsee commented:
Added cleanup of the existing tables, repopulating one entry per stuck threads from f_threads.

nyetwurk commented on db/migrations/20221114170158_revise_sticky_triggers.php:
```diff
@@ -0,0 +1,40 @@
+    "end ";
+	echo "Updating trigger for $triggertbl / $tbl\n";
+  db_exec($sqlTrigger);
+  $sqlTruncateStickyTable = "truncate table  " . $tbl . ";";
+  echo "Truncating table $tbl\n";
+  db_exec($sqlTruncateStickyTable);
+  $sqlUpdateStickyTable = "insert into " . $tbl . " (tid, mid) " .
+  "select tid, mid " .
+  "from " . $triggertbl . " " .
+  "where flags like '%STICKY%';";
+  echo "Backfilling $tbl with stuck threads\n";
+  db_exec($sqlUpdateStickyTable);
+    }
+    $sth->closeCursor();
+  }
```
Wouldn't it be easier (not to mention faster) to just remove duplicates?

nyetwurk commented on db/migrations/20221114170158_revise_sticky_triggers.php:
```diff
@@ -0,0 +1,40 @@
+
+class ReviseStickyTriggers extends DatabaseMigration {
+  public function migrate() {
+    $sth = db_query("select iid from f_indexes order by iid");
+    echo "DO NOT INTERRUPT, this could take quite some time!\n";
+    while ($i = $sth->fetch() ) {
```
this seems super slow.

Can't we change the way this works entirely?

that is to say
1) remove dupes from the sticky tables
2) change the schema to make mid/tid unique

would the existing trigger still be ok?

complacentsee replied:
1. Sure that's probably easier. We can do this:
```
$sqlCleanStickyTable = "DELETE t1 FROM . " .$triggertbl . "t1 " .
  "INNER JOIN " .$triggertbl . "t2 " .
  "WHERE t1.sid < t2.sid AND t1.tid = t2.tid;"
echo "Cleaning table $tbl\n";
db_exec($sqlCleanStickyTable);
```

2. Updating the schema for sticky would work to add a unique constraint, but the trigger would need to be updated to avoid duplicate exceptions. I would want to remove the mid column if we were updating the f_sticky schema. The trigger should get updated to avoid any opportunities of duplicate key exceptions. 
```
  $sqlTrigger = "drop trigger if exists trigger_" . $tbl . "_sticky_update; " .
    "drop trigger if exists trigger_f" . $i['iid'] . "_sticky_update; " .
    "create trigger trigger_f" . $i['iid'] . "_sticky_update " .
    "after update on " . $triggertbl . " " .
    "for each row " .
    "begin " .
    "if new.flags like '%STICKY%' then " .
    "if new.flags <> old.flags then " .
    "insert into " . $tbl . "(tid) values (new.tid) " .
    "on duplicate key update tid=tid;" .
    "end if; " .
    "else " .
    "delete from " . $tbl . " where tid = new.tid; " .
    "end if; " .
    "end ";
```
But if were adding the unique constraint the trigger wouldn't need to confirm old flag value since we can just update the row setting tid = tid and could be simplified to:
```
  $sqlTrigger = "drop trigger if exists trigger_" . $tbl . "_sticky_update; " .
    "drop trigger if exists trigger_f" . $i['iid'] . "_sticky_update; " .
    "create trigger trigger_f" . $i['iid'] . "_sticky_update " .
    "after update on " . $triggertbl . " " .
    "for each row " .
    "begin " .
    "if new.flags like '%STICKY%' then " .
    "insert into " . $tbl . "(tid) values (new.tid) " .
    "on duplicate key update tid=tid;" .
    "else " .
    "delete from " . $tbl . " where tid = new.tid; " .
    "end if; " .
    "end ";
```

nyetwurk replied:
i was thinking more literally, just remove the duplicate entries from the f_sticky tables directly rather than rebuilding them (let alone have the trigger do it).

nyetwurk replied:
Sorry what i couldn't figure out is the relationship between `$triggertbl` and `f_stickyX` - is it *literally* the name of the sticky table?

complacentsee replied:
Yes, that's the relation. I think a cleaner version would look like the following. It's a bit sloppy on the $triggername since I had a typo on the original forum creation script compared to the migration. 
```
class ReviseStickyTriggers extends DatabaseMigration {
  public function migrate() {
    $sth = db_query("select iid from f_indexes order by iid");
    $forumindexids = $sth->fetchAll();
    $sth->closeCursor();
    echo "DO NOT INTERRUPT, this update should complete shortly\n";
    foreach($forumindexids as $forumindexid) {
      $triggertbl = "f_threads" . $forumindexid['iid'];
      $stickytbl = "f_sticky" . $forumindexid['iid'];
      $tiggername = "trigger_f" . $forumindexid['iid'];

      echo "Updating $tiggername for $triggertbl / $stickytbl\n";
      $sqlTriggerCreation = 
        "drop trigger if exists trigger_" . $stickytbl . "_sticky_update; " . //cleans up old trigger created by migration
        "drop trigger if exists " . $tiggername . "_sticky_update; " . //cleans up old trigger created by new forum
        "create trigger " . $tiggername . "_sticky_update " .
        "after update on " . $triggertbl . " " .
        "for each row " .
        "begin " .
        "if new.flags like '%STICKY%' then " .
        "if new.flags <> old.flags then " .
        "insert into " . $stickytbl . "(tid, mid) values (new.tid, new.mid) " .
        "on duplicate key update tid=tid; " .
        "end if; " .
        "else " .
        "delete from " . $stickytbl . " where tid = new.tid; " .
        "end if; " .
        "end ";
      db_exec($sqlTriggerCreation);

      echo "Cleaning douplicates from sticky table $stickytbl\n";
      $sqlCleanStickyTable = "DELETE t1 FROM . " .$stickytbl . " t1 " .
        "INNER JOIN " .$stickytbl . " t2 " .
        "WHERE t1.sid < t2.sid AND t1.tid = t2.tid;";
      db_exec($sqlCleanStickyTable);
    }
  }
}
```

nyetwurk commented:
Is this ready to go?
Do we want to add uniq to the table schema?
Do we want to add a "sticky table dup cleanup" in nightly?

complacentsee commented:
> Is this ready to go?

Updated the commit with the remaining bug fixes associated with the sticky table changes. 
Fixed so that a forum will display with only stuck thread(s). Previously it would display there were no message in the forum if there were only stuck threads. 
> Do we want to add uniq to the table schema?

Added schema changes to f_sticky tables. 
> Do we want to add a "sticky table dup cleanup" in nightly?

I'm not sure this is needed. By updating the schema to avoid duplicate thread ids and updating the trigger to just update records if for some reason it tries to add a duplicate this should be safe. With the distinct selection on the stuck thread count the selection should always be safe even with duplicates, and the duplicates would always be cleaned once a thread is unstuck.

nyetwurk commented:
```1064 You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near '. f_sticky1 t1 INNER JOIN f_sticky1 t2 WHERE t1.sid < t2.sid AND t1.tid = t2.tid' at line 1 in /home/nyet/src/kawf-git/include/sql.inc:73```

nyetwurk commented:
Lots of other problems.

Fixed that syntax error.

Now problem that you add uniq BEFORE removing dups
